File: mcp-server/internal/domains/workflows/apply.go
package workflows

import (
	"context"
	"fmt"

	"github.com/stigmer/stigmer/mcp-server/internal/domains"
	workflowv1 "github.com/stigmer/stigmer/mcp-server/proto/ai/stigmer/agentic/workflow/v1"
	"google.golang.org/grpc"
)

// Apply creates or updates a workflow via the WorkflowCommandController.Apply RPC.
func Apply(ctx context.Context, serverAddress string, workflow *workflowv1.Workflow) (string, error) {
	return domains.WithConnection(ctx, serverAddress,
		func(ctx context.Context, conn *grpc.ClientConn) (string, error) {
			client := workflowv1.NewWorkflowCommandControllerClient(conn)
			result, err := client.Apply(ctx, workflow)
			if err != nil {
				desc := fmt.Sprintf("workflow %q in org %q", workflow.GetMetadata().GetSlug(), workflow.GetMetadata().GetOrg())
				return "", domains.RPCError(err, desc)
			}
			return domains.MarshalJSON(result)
		})
}

File: mcp-server/src/domains/workflows/apply.ts
// Workflow apply path: create-or-update via the WorkflowCommandController.apply
// RPC. The flat MCP input is projected into a fully-formed Workflow proto by the
// generated workflowInputToProto bridge (codegen, src/gen/workflow.ts), which
// expands each task's typed config into the task_config Struct.
//
// Go parity note: the Go MCP server keeps apply_workflow registered but its
// generated input cannot express recursive task nesting in Go's jsonschema
// reflection. protobuf-es + zod's z.lazy resolve that limitation, so the TS
// server exposes apply_workflow at full fidelity.

import { WorkflowSchema } from "@stigmer/protos/ai/stigmer/agentic/workflow/v1/api_pb";
import { WorkflowCommandController } from "@stigmer/protos/ai/stigmer/agentic/workflow/v1/command_pb";

import { ApiResourceVisibility } from "@stigmer/protos/ai/stigmer/commons/apiresource/enum_pb";

import { workflowInputToProto, type WorkflowInput } from "../../gen/workflow.js";
import { applyDeclaredVisibility } from "../apply-visibility.js";
import { withClient } from "../client.js";
import { toProtoJson } from "../marshal.js";
import { rpcError } from "../rpcerr.js";

/** Create or update a workflow, returning the persisted workflow as protojson. */
export async function applyWorkflow(
  serverAddress: string,
  token: string,
  input: WorkflowInput,
): Promise<string> {
  const workflow = workflowInputToProto(input);
  const desc = `workflow "${workflow.metadata?.slug ?? ""}" in org "${workflow.metadata?.org ?? ""}"`;
  return withClient(
    WorkflowCommandController,
    serverAddress,
    token,
    async (client, callOptions) => {
      try {
        const applied = await client.apply(workflow, callOptions);
        const result = await applyDeclaredVisibility(
          client,
          callOptions,
          applied,
          workflow.metadata?.visibility ?? ApiResourceVisibility.api_resource_visibility_unspecified,
        );
        return toProtoJson(WorkflowSchema, result);
      } catch (err) {
        throw rpcError(err, desc);
      }
    },
  );
}
